ONSdigital/eq-terraform
# Eq 854 split into separate terraform scripts

### Important 
- Before checking this out and running it, ensure you tear down your existing AWS environment using the current Terraform scripts (on master)

### What is the context of this PR?
Splits the existing Terraform into separate scripts for VPC, routing, database, queue and application. 

The VPC and internet gateway has been split from the routing (subnets, NAT gateways and route tables) as in some environments (pre-prod, production) the VPC already exists (potentially peered to another VPC). An attempt was made to make these conditional in the routing Terraform, but there doesn't seem to be a way when creating a resource to have properties use either a variable or a conditionally created resource. 

Splitting the database, queue and application allows independent creation and destruction, so we can deploy application infrastructure changes without tearing down the database or queues.

There are a number of networking changes:
- There is now provision for 3 public, 3 application and 3 database subnets split across 3 availability zones. The number of subnets created for each is dependent on the number of CIDR blocks given at build time. This allows a reduced developer build in only one availability zone.
- There is a queue subnet which has routes propagated from the Virtual Private Gateway for connections from SDX. This subnet hasn't been split into two as there is a requirement for the queues to keep their existing static IPs (.15 and .16) requiring the subnet range to span these IPs.
- The public subnets have a NAT gateway in each, these are used to access the internet during elastic beanstalk builds
- The application has been moved into private subnets, accessed from load balancers in public subnets
- The Terraform now also turns on route propagation on the queue route table so that SDX connectivity will work without manual intervention after a build

### How to review
- Check out and build - `export AWS_ENVIRONMENT_NAME=<name>; ./survey-runner.sh plan/apply/destroy`
- Deploy the application and verify that it works, including submission to the queues
